src/main/java/org/opensearch/flowframework/util/EncryptorUtils.java

```diff
@@ -248,99 +285,146 @@ public Template redactTemplateSecuredFields(User user, Template template) {
-            // Using the master_key string as the document id
-            GetRequest getRequest = new GetRequest(CONFIG_INDEX).id(MASTER_KEY);
-            client.get(getRequest, ActionListener.wrap(getResponse -> {
-                if (!getResponse.isExists()) {
-                    Config config = new Config(generateMasterKey(), Instant.now());
-                    IndexRequest masterKeyIndexRequest = new IndexRequest(CONFIG_INDEX).id(MASTER_KEY)
-                        .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
-                    try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
-                        masterKeyIndexRequest.source(config.toXContent(builder, ToXContent.EMPTY_PARAMS));
-                    }
-                    client.index(masterKeyIndexRequest, ActionListener.wrap(indexResponse -> {
-                        context.restore();
-                        // Set generated key to master
-                        logger.info("Config has been initialized successfully");
-                        this.masterKey = config.masterKey();
-                        listener.onResponse(true);
-                    }, indexException -> {
-                        logger.error("Failed to index config", indexException);
-                        listener.onFailure(indexException);
-                    }));
-
+    public void initializeMasterKey(@Nullable String tenantId, ActionListener<Boolean> listener) {
+        // Config index has already been created or verified
+        cacheMasterKeyFromConfigIndex(tenantId).thenApply(v -> {
+            // Key exists and has been cached successfully
+            listener.onResponse(true);
+            return null;
+        }).exceptionally(throwable -> {
+            Exception exception = SdkClientUtils.unwrapAndConvertToException(throwable);
+            // The cacheMasterKey method only completes exceptionally with FFE
+            if (exception instanceof FlowFrameworkException) {
+                FlowFrameworkException ffe = (FlowFrameworkException) exception;
+                if (ffe.status() == RestStatus.NOT_FOUND) {
+                    // Key doesn't exist, need to generate and index a new one
+                    generateAndIndexNewMasterKey(tenantId, listener);
                 } else {
+                    listener.onFailure(ffe);
+                }
+            } else {
+                // Shouldn't get here
+                listener.onFailure(exception);
+            }
+            return null;
+        });
+    }
+
+    private void generateAndIndexNewMasterKey(String tenantId, ActionListener<Boolean> listener) {
+        String masterKeyId = tenantId == null ? MASTER_KEY : MASTER_KEY + "_" + hashString(tenantId);
+        Config config = new Config(generateMasterKey(), Instant.now());
+        PutDataObjectRequest putRequest = PutDataObjectRequest.builder()
+            .index(CONFIG_INDEX)
+            .id(masterKeyId)
+            .tenantId(tenantId)
```

Comment:
There's no issue here when tenantID is null? we are passing the original tenantId which could be null in

Reply:
if tenant awareness isn't enabled, null is the expected value.  The SdkClient validates that tenant ID is non null if multitenancy is enabled, so it's required in that case even if not used otherwise.